File: samples/sample.scala
// ====================================================================
// Scala Syntax Highlighting Sample — NOT compilable, covers many
// syntax constructs for theme testing purposes.
// ====================================================================

// === PACKAGE & IMPORTS ===

package io.github.e1turin.samples

import scala.collection.immutable.{ListMap, VectorMap}
import scala.math.{Pi => π, pow, sqrt}
import scala.util._
import scala.concurrent.{Future, Promise}
import scala.concurrent.ExecutionContext.Implicits.global

// === VAL / VAR / LAZY VAL / TYPE ===

val MaxSize: Int = 1000
val PiConst: Double = 3.14159
val Debug: Boolean = true
val DefaultName: String = "Unnamed"
private var counter: Int = 0
var mutableFlag: Boolean = false

lazy val expensiveResource: String = {
  println("initialising lazy val…")
  "loaded"
}

type ShapeFactory = (String, Double*) => Shape
type Points = Seq[(Double, Double)]

// === DEF — functions, default params, currying, by-name ===

def add(a: Int, b: Int): Int = a + b

def greet(name: String, greeting: String = "Hello"): String =
  s"$greeting, $name!"

def namedArgsExample(): Unit =
  println(greet(greeting = "Hi", name = "Bob"))

def curried(a: Int)(b: Int)(c: Int): Int = a + b + c

def byName(cond: Boolean)(block: => Unit): Unit =
  if (cond) block

def varargs(values: Int*): Int = values.sum

// === IMPLICIT ===

implicit val defaultGreeting: String = "Hello"

def greetImplicit(name: String)(implicit greeting: String): String =
  s"$greeting, $name!"

implicit class RichInt(val x: Int) extends AnyVal {
  def squared: Int = x * x
  def isEven: Boolean = x % 2 == 0
}

implicit def intToOption(x: Int): Option[Int] = Some(x)

// === TRAIT ===

trait Drawable {
  def draw(): String
  def description: String = "A drawable object"
}

trait Movable {
  def move(dx: Double, dy: Double): Unit
}

sealed trait Status
sealed abstract class Event

// === CLASS — constructor params, extends, with ===

class Circle(val radius: Double, val name: String = "Circle")
    extends Shape
    with Drawable {
  require(radius > 0, "Radius must be positive")

  override def area(): Double = π * radius * radius
  override def perimeter(): Double = 2 * π * radius
  override def draw(): String = s"Circle($radius)"
}

class Rectangle(
    val width: Double,
    val height: Double,
    private var _name: String = "Rectangle"
) extends Shape
    with Drawable {
  override def area(): Double = width * height
  override def perimeter(): Double = 2 * (width + height)
  override def draw(): String = s"Rectangle($width × $height)"
}

// === CASE CLASS / CASE OBJECT ===

case class Point(x: Double, y: Double) {
  def distanceTo(other: Point): Double =
    sqrt((x - other.x) * (x - other.x) + (y - other.y) * (y - other.y))
}

case object Origin extends Point(0, 0)

case class Error(code: Int, message: String) extends Status

// === OBJECT / COMPANION OBJECT / APPLY / UNAPPLY ===

object ShapeRegistry {
  private val shapes = scala.collection.mutable.ListBuffer.empty[Shape]

  def register(shape: Shape): Unit = shapes += shape
  def all(): List[Shape] = shapes.toList
  def clear(): Unit = shapes.clear()
}

object User {
  private var nextId = 1L

  def apply(name: String, email: Option[String] = None): User = {
    val id = nextId; nextId += 1; new User(id, name, email)
  }

  def unapply(u: User): Option[(Long, String, Option[String])] =
    Some((u.id, u.name, u.email))
}

class User private (val id: Long, val name: String, val email: Option[String]) {
  override def toString: String = s"User(id=$id, name='$name')"
}

// === ENUM — Scala 2 Enumeration + Scala 3 enum ===

object Color extends Enumeration {
  type Color = Value
  val Red: Color   = Value("#FF0000")
  val Green: Color = Value("#00FF00")
  val Blue: Color  = Value("#0000FF")
}

// Scala 3 enum (commented for Scala 2 compat; syntax reference)
// enum TrafficLight:
//   case Red, Yellow, Green
//   def next: TrafficLight = this match
//     case Red    => Green
//     case Yellow => Red
//     case Green  => Yellow

// === PATTERN MATCHING ===

val describe: Status => String = {
  case Active           => "Active"
  case Inactive         => "Inactive"
  case Pending          => "Pending…"
  case Error(code, msg) => s"Error $code: $msg"
}

def matchDemo(s: Shape): String = s match {
  case c: Circle    => s"Circle(r=${c.radius})"
  case r: Rectangle => s"Rectangle(${r.width}×${r.height})"
  case _            => "Unknown shape"
}

def guardDemo(x: Int): String = x match {
  case n if n < 0  => "negative"
  case n if n == 0 => "zero"
  case n           => s"positive ($n)"
}

// === PARTIAL FUNCTION ===

val divide: PartialFunction[Int, Int] = {
  case d: Int if d != 0 => 42 / d
}

val safeSqrt: PartialFunction[Double, Double] = {
  case x if x >= 0 => sqrt(x)
}

// === FOR COMPREHENSION / FOR LOOP ===

val coordinates: List[(Int, Int)] = for {
  x <- List(1, 2, 3)
  y <- List(4, 5, 6)
  if x < y
} yield (x, y)

val optResult: Option[Int] = for {
  a <- Some(10)
  b <- Some(20)
} yield a + b

for {
  i <- 1 to 10
  if i % 2 == 0
  if i > 3
} println(s"Even > 3: $i")

// === IF / ELSE / WHILE / DO WHILE ===

def signum(x: Int): String =
  if (x > 0) "positive"
  else if (x < 0) "negative"
  else "zero"

var i = 0
while (i < 3) { println(i); i += 1 }

var j = 0
do { println(j); j += 1 } while (j < 3)

// === COLLECTIONS ===

val list: List[Int]          = List(1, 2, 3, 4, 5)
val vector: Vector[Int]      = Vector(1, 2, 3)
val set: Set[Int]           = Set(1, 2, 3, 1)
val map: Map[String, Int]   = Map("one" -> 1, "two" -> 2, "three" -> 3)
val seq: Seq[Int]           = Seq(1, 2, 3)
val indexed: IndexedSeq[Int] = IndexedSeq(1, 2, 3)
val range: Range             = 1 to 10 by 2

// === HIGHER-ORDER FUNCTIONS ===

val doubled: List[Int]        = list.map(_ * 2)
val evens: List[Int]          = list.filter(_ % 2 == 0)
val flatMapped: List[Int]    = list.flatMap(x => List(x, x * 10))
val sum: Int                  = list.reduce(_ + _)
val foldSum: Int             = list.foldLeft(0)(_ + _)
val product: Int             = list.foldRight(1)(_ * _)
val collected: List[Int]     = list.collect { case n if n > 2 => n * n }
val (big, small)             = list.partition(_ > 3)
val zipped: List[(Int, Int)] = list.zip(List(5, 4, 3, 2, 1))
val flattened: List[Int]     = List(List(1, 2), List(3, 4)).flatten
val grouped: Map[Boolean, List[Int]] = list.groupBy(_ % 2 == 0)

// === OPTION / EITHER / TRY ===

val someVal: Option[Int] = Some(42)
val noneVal: Option[Int] = None

val eitherRight: Either[String, Int] = Right(42)
val eitherLeft: Either[String, Int]  = Left("error")

val trySuccess: Try[Int] = Success(42)
val tryFailure: Try[Int] = Failure(new Exception("boom"))

// === FUTURE / PROMISE ===

val future: Future[Int] = Future { 42 }
val promise: Promise[Int] = Promise[Int]()
val promisedFuture: Future[Int] = promise.future

// === STRING INTERPOLATION & LITERALS ===

val lang      = "Scala"
val greeting  = s"Hello, $lang! — 2 + 2 = ${2 + 2}"
val formatted = f"Pi is ~${π}%.2f and e is ~${2.71828}%.3f"
val rawStr    = raw"Hello\nWorld"     // \n not escaped
val multi     =
  """This is a
    |multiline string
    |with .stripMargin""".stripMargin

// === LITERALS — numeric, char, symbol, null ===

val intLit: Int       = 42
val longLit: Long     = 123L
val floatLit: Float   = 3.14f
val doubleLit: Double = 3.14159
val hexLit: Int       = 0xFF
val binLit: Int       = 0b1010
val charLit: Char     = 'A'
val boolLit: Boolean  = true
val symLit: Symbol    = 'mySymbol
val nullLit: String   = null

// === TUPLES & DESTRUCTURING ===

val t: (String, Int, Boolean) = ("Alice", 30, true)
val (name, age, isAdmin) = t
val pair: (Int, String) = 1 -> "one"

// === GENERICS — variance, bounds, context bounds ===

class Box[+T](val value: T)                     // covariance
class Container[-T](val dummy: Unit)             // contravariance

def identical[A](a: A): A = a

def upperBound[T <: Comparable[T]](a: T, b: T): Int = a.compareTo(b)
def lowerBound[T >: Null](t: T): String = if (t == null) "null" else "not null"

def sortWithOrdering[T: Ordering](items: List[T]): List[T] =
  items.sorted

// === ABSTRACT / OVERRIDE / FINAL / SEALED ===

abstract class Base {
  def mustImplement: Int
  final def fixed: String = "immutable"
}

class Derived extends Base {
  override def mustImplement: Int = 42
}

sealed trait Direction
case object North extends Direction
case object South extends Direction
case object East  extends Direction
case object West  extends Direction

// === ANNOTATIONS ===

@deprecated("Use newMethod instead", "3.0")
def oldMethod(): Unit = println("legacy")

@scala.annotation.tailrec
def gcd(a: Int, b: Int): Int =
  if (b == 0) a else gcd(b, a % b)

@inline
def fastDouble(x: Int): Int = x * 2

// === REQUIRE / ASSERT ===

def positiveRadius(r: Double): Double = {
  require(r > 0, s"radius $r must be positive")
  r
}

def testAssert(x: Int): Unit = {
  assert(x >= 0, "x must be non-negative")
}

// === THROW / TRY / CATCH / FINALLY ===

def riskyDivide(a: Int, b: Int): Int =
  if (b == 0) throw new IllegalArgumentException("division by zero")
  else a / b

def safeDivide(a: Int, b: Int): Try[Int] = Try {
  if (b == 0) throw new ArithmeticException("/ by zero")
  a / b
} recover {
  case _: ArithmeticException => 0
}

def withFinally(): Unit = {
  val f = scala.io.Source.fromFile("/etc/hostname")
  try {
    println(f.getLines().mkString)
  } catch {
  case e: java.io.IOException => println(s"I/O error: $e")
  } finally {
    f.close()
  }
}

// === NEW ===

val freshCircle = new Circle(2.5, "tiny")

// === SELF TYPE ===

trait Persistence { def save(id: String, data: String): Unit }

trait Logger { self: Persistence =>
  def logAndSave(id: String, data: String): Unit = {
    println(s"Logging $id")
    save(id, data)
  }
}

// === SCALA 3: GIVEN / USING (illustrative) ===

// Scala 3 syntax (commented for Scala 2 compat)
// given Ordering[Int] with
//   def compare(a: Int, b: Int): Int = a - b
//
// def max[A](a: A, b: A)(using ord: Ordering[A]): A =
//   if (ord.compare(a, b) >= 0) a else b

// === SCALA 3: EXTENSION METHODS ===

// extension (x: Int)
//   def cubed: Int = x * x * x

// === SCALA 3: EXPORT ===

// class Exporter(c: Circle):
//   export c.{area, perimeter => perim}
//   export c.description

// === SAMPLE VALUES (to avoid dead-code warnings in some tools) ===

val sampleValues: List[Any] = List(
  freshCircle, sum, foldSum, product, doubled, evens,
  future, someVal, eitherRight, trySuccess, coordinates,
  range, flattened, grouped, big, small, zipped,
  collected, 42.squared, 4.isEven, defaultGreeting
)

File: samples/sample.kt
// Kotlin sample for Alabaster theme
// Demonstrates common Kotlin language features

package io.github.e1turin.samples

import kotlin.math.PI
import kotlin.math.pow
import kotlin.system.measureTimeMillis

// Global constants
const val MAX_SIZE: Int = 1000
const val PI_CONST: Double = 3.14159
const val DEBUG: Boolean = true
const val DEFAULT_NAME: String = "Unnamed"

// Top-level property
val counter: Int = 0
var mutableCounter = 0

// Type alias
typealias ShapeFactory = (String, DoubleArray) -> Shape?

// Data class
data class Point(val x: Double, val y: Double) {
    fun distanceTo(other: Point): Double {
        val dx = x - other.x
        val dy = y - other.y
        return kotlin.math.sqrt(dx * dx + dy * dy)
    }
}

// Enum class with properties
enum class Status {
    ACTIVE,
    INACTIVE,
    PENDING,
    ERROR
}

// Enum with constructor and methods
enum class Color(val hex: String) {
    RED("#FF0000"),
    GREEN("#00FF00"),
    BLUE("#0000FF");

    fun rgb(): Triple<Int, Int, Int> {
        val r = hex.substring(1..2).toInt(16)
        val g = hex.substring(3..4).toInt(16)
        val b = hex.substring(5..6).toInt(16)
        return Triple(r, g, b)
    }
}

// Sealed class / interface
sealed class Shape {
    abstract fun area(): Double
    abstract fun perimeter(): Double
}

// Inheritance: class extending sealed class
class Circle(val radius: Double, val name: String = "Circle") : Shape() {
    init {
        require(radius > 0) { "Radius must be positive" }
    }

    override fun area(): Double = PI * radius.pow(2)

    override fun perimeter(): Double = 2 * PI * radius
}

class Rectangle(val width: Double, val height: Double, name: String = "Rectangle") : Shape() {
    override fun area(): Double = width * height

    override fun perimeter(): Double = 2 * (width + height)
}

// Object declaration (singleton)
object ShapeRegistry {
    private val shapes = mutableListOf<Shape>()

    fun register(shape: Shape) {
        shapes.add(shape)
    }

    fun all(): List<Shape> = shapes.toList()

    fun clear() {
        shapes.clear()
    }
}

// Companion object and factory pattern
class User private constructor(
    val id: Long,
    val name: String,
    val email: String?
) {
    companion object Factory {
        private var nextId = 1L

        @JvmStatic
        fun create(name: String, email: String? = null): User {
            return User(nextId++, name, email)
        }
    }

    val displayName: String
        get() = email?.let { "$name <$it>" } ?: name

    override fun toString(): String = "User(id=$id, name='$name')"
}

// Interface with default implementation
interface Drawable {
    fun draw(): String

    fun description(): String = "A drawable object"
}

// Multiple interface implementation
class Drawing(private val shapes: List<Shape>) : Drawable {
    override fun draw(): String {
        return shapes.joinToString("\n") { shape ->
            when (shape) {
                is Circle -> "Circle(radius=${shape.radius})"
                is Rectangle -> "Rectangle(w=${shape.width}, h=${shape.height})"
                else -> "Unknown shape"
            }
        }
    }
}

// Extension functions
fun Circle.scaled(factor: Double): Circle {
    return Circle(radius * factor, "$name (scaled)")
}

fun String.isEmail(): Boolean {
    return this.contains("@") && this.contains(".")
}

// Generic function
fun <T> identity(value: T): T = value

// Higher-order function with lambda
fun <T> List<T>.customFilter(predicate: (T) -> Boolean): List<T> {
    val result = mutableListOf<T>()
    for (item in this) {
        if (predicate(item)) {
            result.add(item)
        }
    }
    return result
}

// Inline function
inline fun <reified T> isInstance(value: Any): Boolean {
    return value is T
}

// Suspend function (coroutines demonstration)
suspend fun fetchData(url: String): String {
    // Simulated async work
    kotlinx.coroutines.delay(100)
    return "Data from $url"
}

// Main function
fun main() {
    // Various string types
    val singleQuoted = "Hello, World!"
    val rawString = """Raw
        |multiline
        |string literal""".trimMargin()

    // String templates
    val name = "Kotlin"
    val greeting = "Hello, $name!"
    val expression = "2 + 2 = ${2 + 2}"

    // Numeric literals
    val integer = 42
    val float = 3.14
    val hex = 0xFF
    val binary = 0b1010
    val long = 123L
    val unsigned = 42u

    // Boolean literals
    val isTrue = true
    val isFalse = false

    // Nullable types
    val nullableString: String? = "not null"
    val nullString: String? = null

    // Null safety
    val length: Int? = nullableString?.length
    val safeLength: Int = nullString?.length ?: -1

    // Creating instances
    val circle = Circle(5.0, "My Circle")
    val rect = Rectangle(10.0, 20.0)
    val point = Point(3.0, 4.0)

    // Using objects
    ShapeRegistry.register(circle)
    ShapeRegistry.register(rect)

    // For loop with range
    for (i in 1..10) {
        if (i % 2 == 0) continue
        if (i > 7) break
        println(i)
    }

    // For loop with collection
    val items = listOf("a", "b", "c")
    for ((index, value) in items.withIndex()) {
        println("$index: $value")
    }

    // While loop
    var i = 0
    while (i < 3) {
        println(i)
        i++
    }

    // When expression (replacement for switch)
    val status = Status.ACTIVE
    val message = when (status) {
        Status.ACTIVE -> "Active"
        Status.INACTIVE -> "Inactive"
        Status.PENDING -> "Pending..."
        Status.ERROR -> "Error occurred"
    }

    // Collections
    val list = listOf(1, 2, 3, 4, 5)
    val mutableList = mutableListOf(1, 2, 3)
    val set = setOf(1, 2, 3)
    val map = mapOf("key1" to "value1", "key2" to "value2")

    // Destructuring declarations
    val (x, y) = Point(10.0, 20.0)
    println("Point: x=$x, y=$y")

    // Scope functions
    val result = circle.let {
        "Area: ${it.area()}, Perimeter: ${it.perimeter()}"
    }

    val rectCopy = rect.apply {
        // apply provides 'this' context
        println("Using apply on $this")
    }

    val area = with(rect) {
        area() // same as rect.area()
    }

    // Lambda usage
    val doubled = list.map { it * 2 }
    val evens = list.filter { it % 2 == 0 }
    val sum = list.reduce { acc, n -> acc + n }

    // Generic function
    val intIdentity = identity(42)
    val stringIdentity = identity("hello")

    // Extension function usage
    val scaledCircle = circle.scaled(2.0)
    val isValidEmail = "user@example.com".isEmail()

    // Companion object usage
    val user = User.create("Alice", "alice@example.com")

    // Annotations
    @Suppress("UNUSED_EXPRESSION")
    val unused = 42
}
